Add: Bulk actions to dataviews with the new design.

jorgefilipecosta opened the pull request:
Supersedes: https://github.com/WordPress/gutenberg/pull/56476, https://github.com/WordPress/gutenberg/pull/56615

This PR proposes a bulk actions implementation using the DropDownMenuV2 component trying to follow the design proposed at https://github.com/WordPress/gutenberg/pull/56476#issuecomment-1845606302.

For now, bulk actions are implemented only for templates.
If multiple items are selected, each one supporting different bulk actions, we show the number of elements the action applies to and only apply the action to the eligible elements.

The actions API was expanded with a supportsBulk option. We need to know if the action supports a bulk option or one to decide if it appears in the bulk edit menu. E.g.: A see revisions action would not make sense under bulk edit.

The actions API callback and Render modal were also changed so the items are an array (support multiple items instead of just one).

## Screenshots or screencast <!-- if applicable -->

<img width="1407" alt="Screenshot 2023-12-20 at 10 56 51" src="https://github.com/WordPress/gutenberg/assets/11271197/53ecdfae-421a-4e11-91d9-85b465dbef73">

jorgefilipecosta committed:
Add: Bulk actions to dataviews with the new design.

jameskoster commented:
I think the work here can pave the way for more elaborate designs in the future, but for now this feels like a good starting point that is lightweight and accessible. 

Visually the menu doesn't seem to match the latest [DropdownMenuV2](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-dropdownmenu-v2--docs) designs, perhaps that will be fixed by a rebase?

Could we try moving the 'Bulk edit' button to the right, just left of the 'View options' button?

> For now, bulk actions are implemented only for templates.

I'm seeing the selection controls on the Pages list, should those be disabled altogether if there are no actions?

youknowriad commented:
I was kind of surprised to see the checkboxes and bulk actions kind of visible by default. I think I've expected to have to first click "bulk edit" to show the checkboxes and the bulk actions menu. Just wondering if this is something that has been discussed?

youknowriad commented:
In the "pages" dataviews, I can see the checkboxes and the bulk selection menu even though I don't see any bulk actions to perform.

youknowriad commented:
Probably a follow-up but did we think about the UI for bulk selection in "list" and "grid" layouts?

youknowriad commented:
Maybe I'm doing something wrong but clicking the "reset" or "delete template" bulk actions does nothing for me.

jorgefilipecosta committed:
post rebase fixes.

jorgefilipecosta commented:
> Maybe I'm doing something wrong but clicking the "reset" or "delete template" bulk actions does nothing for me.

Sorry, I missed pushing a post-rebase fix I had locally (onSelect -> onClick on a component that changed) it should be working now.

jorgefilipecosta commented:
> Probably a follow-up but did we think about the UI for bulk selection in "list" and "grid" layouts?

I will implement the other views bulk actions in separate PRs to keep this one smaller. As far as I know, there are no mockups yet. I expect both for list and grid pressing shift plus a click would trigger multi selection of elements that seems to be a common pattern. But I'm to other options.

jorgefilipecosta commented:
> I was kind of surprised to see the checkboxes and bulk actions kind of visible by default. I think I've expected to have to first click "bulk edit" to show the checkboxes and the bulk actions menu. Just wondering if this is something that has been discussed?

"I believe we haven't discussed this option yet. The Bulk Edit menu triggers a dropdown, which could make it a bit confusing to display checkboxes as well. To use this option, we would first need to click on "Bulk edit", which would display the checkboxes along with an actions menu. When we click on the actions menu, the current dropdown will appear with the actions and options to select all or deselect all. We would also need a button like stop bulk edit to hide checkboxes again.

We can also consider another option where we hide the checkboxes to reduce visual clutter, but display them when we hover the mouse over a row.

jorgefilipecosta committed:
Add missing secondary variant

jorgefilipecosta committed:
Move bulk actions button position

jorgefilipecosta committed:
Don't render bulk actions functionality when there are no bulk actions avaliable.

jorgefilipecosta commented:
> I think the work here can pave the way for more elaborate designs in the future, but for now this feels like a good starting point that is lightweight and accessible.
> 
> Visually the menu doesn't seem to match the latest [DropdownMenuV2](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-dropdownmenu-v2--docs) designs, perhaps that will be fixed by a rebase?
> 
> Could we try moving the 'Bulk edit' button to the right, just left of the 'View options' button?

I added a secondary style to the bulk edit button, it should now have a similar style to the story book sample.
I also changed the position as suggested.


> > For now, bulk actions are implemented only for templates.
> 
> I'm seeing the selection controls on the Pages list, should those be disabled altogether if there are no actions?

The issue was fixed.

jorgefilipecosta commented:
> In the "pages" dataviews, I can see the checkboxes and the bulk selection menu even though I don't see any bulk actions to perform.

Fixed 👍

jameskoster commented:
> Just wondering if this is something that has been discussed?

A 'bulk edit' mode was explored early on (similar to the media library in wp-admin) but it didn't feel super compelling, mostly due to the friction it adds to the selection process. It was quite difficult to answer: "why not make those controls permanently available"? It also raised questions about other table features when such a mode was active, e.g. would it still be possible to perform inline edits/actions, filtering, etc? In short it solves some issues while introducing others.

I would say that the checkboxes themselves feel quite prominent visually, and that perhaps overemphasises bulk actions. This could be a detail to visit separately, and there are a few options to explore there:

* Reduce checkbox visual size (most design systems use ≤16px, checkboxes in Gutenberg are currently 20px).
* Only show checkbox on row hover/selection.
* Include a view option to hide checkbox column. This would effectively allow users to create a custom "bulk edit" view if they wanted to, kind of similar to the 'bulk edit mode' that was explored initially.

youknowriad commented:
>  "why not make those controls permanently available"

I guess for me the answer to this is that I don't see bulk editing as the primary action on the lists, and something you'd rarely don, only for advanced use-cases. So it seems counter intuitive for me for these controls to be so prominent.

youknowriad commented:
<img width="126" alt="Screenshot 2024-01-04 at 4 29 05 PM" src="https://github.com/WordPress/gutenberg/assets/272444/fee155a7-a15c-4042-961d-29e6782a8f5e">

This button is not well spaced.

Also, I noticed that there's a focus loss issue on both actions, maybe a good approach would be to focus the "bulk edit" button once the action is performed.

Other than that, this seems fine to land as v1

jorgefilipecosta committed:
Fix button padding

jorgefilipecosta committed:
Fix focus loss

jorgefilipecosta commented:
> <img alt="Screenshot 2024-01-04 at 4 29 05 PM" width="126" src="https://private-user-images.githubusercontent.com/272444/294224109-fee155a7-a15c-4042-961d-29e6782a8f5e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ0MTA4NjUsIm5iZiI6MTcwNDQxMDU2NSwicGF0aCI6Ii8yNzI0NDQvMjk0MjI0MTA5LWZlZTE1NWE3LWExNWMtNDA0Mi05NjFkLTI5ZTY3ODJhOGY1ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEwNFQyMzIyNDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02Mjc5ZTgxNzdiMmQ1Y2EyODQ1ZDVjY2NkZTk3NThiMTg1MmZlMjhmNzYzNDVkYmQxZDZjNTBmNDMyMzBlNmZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.HBtuY0RnMKx9sJc_jHimch38unPRESTkmAraodFvVvQ">
> This button is not well spaced.
> 
> Also, I noticed that there's a focus loss issue on both actions, maybe a good approach would be to focus the "bulk edit" button once the action is performed.
> 
> Other than that, this seems fine to land as v1

Hi @youknowriad thank you for catching these issues, I think both of them were fixed.

youknowriad commented on packages/dataviews/src/style.scss:
```diff
@@ -287,7 +287,23 @@
 }
+
+.dataviews-bulk-edit-button.components-button.is-secondary {
+	padding-right: $grid-unit-15 + 2px;
```
Why we need a special padding in the first place. For me, the issue seems to be that the flex behavior is "shrinking" the width of the button so instead of a custom padding (which result in a wrong balance for me btw), we need to use `flex-string: 0`

jorgefilipecosta replied:
Good one flex-shrink did the trick. I did not find the root cause of why the button was cut on the right and the padding was like a last resource to address the issue.

youknowriad commented on packages/dataviews/src/bulk-actions.js:
```diff
@@ -21,24 +21,31 @@ const {
+	action,
+	selectedItems,
+	setActionWithModal,
+	bulkActionsButtonRef,
```
I don't like this prop. In general I don't like passing refs and Dom elements as props because: Which can of ref is this, an object, a function?... Also, they're subject to timing changes and mutations.

A better API here could be `onPerform` or something like that. Something to be called when the action is done for instance. 

That said, I do wonder: I believe modals and popover have a built-in behavior to restore the focus to the caller (button that opened the popover/modal), in this case it doesn't work natively, so I'm curious to know why the "focus return" don't work by default when the dropdown get closed. cc @ciampo

ciampo replied:
> I believe modals and popover have a built-in behavior to restore the focus to the caller

The `Modal`, when closed, will try to restore focus on the element that was last focused before opening the modal — ie. the menu item. Similarly, the DropdownMenu will try to restore focus on the menu trigger when closed.

Therefore, if we want to focus the dropdown menu trigger when closing the modal, that may be something that needs to be implemented in a custom way. An alternative that we could consider is to leave the dropdown menu open — in theory, the modal component should be able to restore focus correctly on the menu item when closed (see [storybook example](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-dropdownmenu-v2--with-modals)).

youknowriad replied:
> The Modal, when closed, will try to restore focus on the element that was last focused before opening the modal — ie. the menu item. Similarly, the DropdownMenu will try to restore focus on the menu trigger when closed.

I guess in my mind, when I click the trigger it should get focused and then open the modal so it's the last item that was focused before the modal is opened. 

But now that you bring this up, I remember some discrepancies between browsers and not all of them have the "focus on click" behavior. Which means that maybe it's because of me testing on Safari which makes me wonder whether we should bother about fixing this issue (should we just revert the commit?) as this is the expected behavior there. I'll check in chrome (before the commit) to confirm.

ciampo replied:
I went ahead and tried to implement an alternative solution that doesn't rely on `bulkActionsButtonRef` at all (and keeps the dropdown menu open in the background when rendering the modal). Could this work?

<details>

<summary>Click to expand</summary>

```diff
diff --git a/packages/dataviews/src/bulk-actions.js b/packages/dataviews/src/bulk-actions.js
index 9613ff37ee..3fa5db3d76 100644
--- a/packages/dataviews/src/bulk-actions.js
+++ b/packages/dataviews/src/bulk-actions.js
@@ -25,7 +25,7 @@ function ActionWithModal( {
 	action,
 	selectedItems,
 	setActionWithModal,
-	bulkActionsButtonRef,
+	onMenuOpenChange,
 } ) {
 	const eligibleItems = useMemo( () => {
 		return selectedItems.filter( ( item ) => action.isEligible( item ) );
@@ -33,11 +33,10 @@ function ActionWithModal( {
 	const { RenderModal, hideModalHeader } = action;
 	const onCloseModal = useCallback( () => {
 		setActionWithModal( undefined );
-		// Focus the bulk actions button.
-		if ( bulkActionsButtonRef?.current?.focus ) {
-			bulkActionsButtonRef.current.focus();
-		}
-	}, [ setActionWithModal, bulkActionsButtonRef ] );
+		// Close dropdownmenu
+		// TODO: should be called only if the primary action of the modal is clicked
+		onMenuOpenChange( false );
+	}, [ setActionWithModal, onMenuOpenChange ] );
 	return (
 		<Modal
 			title={ ! hideModalHeader && action.label }
@@ -50,31 +49,23 @@ function ActionWithModal( {
 	);
 }
 
-function BulkActionItem( {
-	action,
-	selectedItems,
-	onMenuOpenChange,
-	setActionWithModal,
-	bulkActionsButtonRef,
-} ) {
+function BulkActionItem( { action, selectedItems, setActionWithModal } ) {
 	const eligibleItems = useMemo( () => {
 		return selectedItems.filter( ( item ) => action.isEligible( item ) );
 	}, [ action, selectedItems ] );
+
+	const shouldShowModal = !! action.RenderModal;
+
 	return (
 		<DropdownMenuItem
 			key={ action.id }
 			disabled={ eligibleItems.length === 0 }
-			onClick={ async ( event ) => {
-				event.preventDefault();
-				if ( !! action.RenderModal ) {
-					onMenuOpenChange( false );
+			hideOnClick={ ! shouldShowModal }
+			onClick={ async () => {
+				if ( shouldShowModal ) {
 					setActionWithModal( action );
 				} else {
 					await action.callback( eligibleItems );
-					// Click the bulk actions button to close the dropdown and focus the button.
-					if ( bulkActionsButtonRef?.current?.click ) {
-						bulkActionsButtonRef.current.click();
-					}
 				}
 			} }
 			suffix={
@@ -86,13 +77,7 @@ function BulkActionItem( {
 	);
 }
 
-function ActionsMenuGroup( {
-	actions,
-	selectedItems,
-	onMenuOpenChange,
-	setActionWithModal,
-	bulkActionsButtonRef,
-} ) {
+function ActionsMenuGroup( { actions, selectedItems, setActionWithModal } ) {
 	return (
 		<>
 			<DropdownMenuGroup>
@@ -101,9 +86,7 @@ function ActionsMenuGroup( {
 						key={ action.id }
 						action={ action }
 						selectedItems={ selectedItems }
-						onMenuOpenChange={ onMenuOpenChange }
 						setActionWithModal={ setActionWithModal }
-						bulkActionsButtonRef={ bulkActionsButtonRef }
 					/>
 				) ) }
 			</DropdownMenuGroup>
@@ -131,7 +114,7 @@ export default function BulkActions( {
 			selection.includes( getItemId( item ) )
 		);
 	}, [ selection, data, getItemId ] );
-	const bulkActionsButtonRef = useRef();
+
 	if ( bulkActions.length === 0 ) {
 		return null;
 	}
@@ -143,7 +126,6 @@ export default function BulkActions( {
 				label={ __( 'Filters' ) }
 				trigger={
 					<Button
-						ref={ bulkActionsButtonRef }
 						className="dataviews-bulk-edit-button"
 						__next40pxDefaultSize
 						variant="secondary"
@@ -164,19 +146,14 @@ export default function BulkActions( {
 			>
 				<ActionsMenuGroup
 					actions={ bulkActions }
-					data={ data }
-					selection={ selection }
-					getItemId={ getItemId }
-					onMenuOpenChange={ onMenuOpenChange }
 					setActionWithModal={ setActionWithModal }
 					selectedItems={ selectedItems }
-					bulkActionsButtonRef={ bulkActionsButtonRef }
 				/>
 				<DropdownMenuGroup>
 					<DropdownMenuItem
 						disabled={ areAllSelected }
-						onClick={ ( event ) => {
-							event.preventDefault();
+						hideOnClick={ false }
+						onClick={ () => {
 							onSelectionChange( data );
 						} }
 						suffix={ data.length }
@@ -185,8 +162,8 @@ export default function BulkActions( {
 					</DropdownMenuItem>
 					<DropdownMenuItem
 						disabled={ selection.length === 0 }
-						onClick={ ( event ) => {
-							event.preventDefault();
+						hideOnClick={ false }
+						onClick={ () => {
 							onSelectionChange( [] );
 						} }
 					>
@@ -199,7 +176,7 @@ export default function BulkActions( {
 					action={ actionWithModal }
 					selectedItems={ selectedItems }
 					setActionWithModal={ setActionWithModal }
-					bulkActionsButtonRef={ bulkActionsButtonRef }
+					onMenuOpenChange={ onMenuOpenChange }
 				/>
 			) }
 		</>

```
</details>

youknowriad replied:
Actually, I just did more testing and here's my findings:

**First issue**

For the "reset template" action, there's no modal, but what's happening there is that the "dropdown" stays open but the action become "disabled" which means there's a focus loss. I believe there are two things we should do here:

 - The simple solution is to just close the dropdown when the action is done `onActionPerformed` or something which should in theory restore the focus to the trigger.
 - The second thing that we need to do IMO (even if it's probably not necessary once the first solution is implemented) is that "disabled" menu items in the dropdown menu shouldn't be "disabled" entirely, they should use the `aria-disabled` trick.  (Equivalent to the `__experimentalIsFocusable` prop on the Button component). This will allow the buttons to stay focusable and prevent focus loss even though the action is disabled.

**Second issue**

For the "delete template" action, there's a modal, so there are nested dialogs basically. And in this case, everything closes and there's a focus loss. I'm not entirely sure what's happening here so I have no solutions to offer for this case yet, but my guess is that with the "aria-disabled" trick that I suggest above, there's a small chance of this issue being fixed as well.

---

I guess what I'm saying is that we shouldn't try to add custom code to solve these focus loss issues for this particular UI, instead we should try first to address the systemic issues that I raise above. Does that make sense?

ciampo replied:
Related to the first issue, the menu items should not lose focus when becoming disabled as [they are already marked as "accessible when disabled" ](https://github.com/WordPress/gutenberg/blob/ae7467d8bab6cbe538ac0de648b57a62480e2bdc/packages/components/src/dropdown-menu-v2/index.tsx#L57). Have you tried inspecting the `document.activeElement` and/or using arrow keys to move the selection to other dropdown menu items? Maybe the issue is only visual (ie. the item doesn't _look_ focused because it's disabled).

Related to the second issue, have you tried the solution that I proposed in the [message above](https://github.com/WordPress/gutenberg/pull/57255#discussion_r1442663814)? IMO we should keep the dropdownmenu open under the modal:

- when pressing "cancel" in the modal, we simply close the modal and the focus gets moved back to the dropdown menu automatically;
- when pressing the primary action (ie. "Delete"), we can programmatically set the DropdownMenu open state to `false`, which should then cause the dropdownmenu to automatically move the focus on its trigger.

youknowriad replied:
> Have you tried inspecting the document.activeElement and/or using arrow keys to move the selection to other dropdown menu items? Maybe the issue is only visual (ie. the item doesn't look focused because it's disabled).

 - It actually becomes a "div" and not a button anymore. Weird. So it's a similar issue but different trigger. so yes, not the aria-disabled trick.
 - The activeElement is the dropdown wrapper in my tests, so it seems there's some code in place already that handle a focus fallback or something.

youknowriad replied:
And for the second issue, you're right as well, the popover is closed when the modal is opened. I'm not entirely sure if we do this on purpose or not but could this be the result of "focus loss". Since the focus is not on the popover anymore, it just closes itself. In other words, our popovers/dropdowns might not support nested dialogs being focused properly yet.

ciampo replied:
> * It actually becomes a "div" and not a button anymore. Weird. So it's a similar issue but different trigger. so yes, not the aria-disabled trick.
> * The activeElement is the dropdown wrapper in my tests, so it seems there's some code in place already that handle a focus fallback or something.

I think all dropdown menu items are actually `div[role="menuitem"]` elements.

Regarding the `activeElement`:
- when using the keyboard to click the "Deselect" item, focus stays on the item
- when using the mouse, focus moves to the menu wrapper

So I believe that focus is already being handled correctly by the underlying ariakit implementation.

> And for the second issue, you're right as well, the popover is closed when the modal is opened. I'm not entirely sure if we do this on purpose or not but could this be the result of "focus loss". Since the focus is not on the popover anymore, it just closes itself.

Closing the dropownmenu when opening the modal is indeed the cause for focus loss. If we kept the dropdownmenu open, the menu item that was clicked to open the modal would receive focus when the modal gets closed.

> In other words, our popovers/dropdowns might not support nested dialogs being focused properly yet.

IMO, the better approach (also UX-wise) is to keep the dropdown menu open under the modal. Doing so should also allow the components to handle focus correctly without extra code:

- Modal would handle focus restoration when dismissing the modal by focusing the correct dropdown menu item
- If we want to force the dropdown menu to close, setting its `open` prop to `false` should move focus automatically on the dropdownmenu trigger. So we should be already able to cover 

When used that way, our components should already handle focus correctly.

youknowriad replied:
> IMO, the better approach (also UX-wise) is to keep the dropdown menu open under the modal

I agree but my question was: Is that even possible with our current components. I believe most of our "popovers"... close themselves when they lose focus to a child modal/popover but you say otherwise, so that's news for me :)

ciampo replied:
Yup! The new `DropdownMenu` menu doesn't use our underlying `Modal` component, and can be definitely be kept open under a `Modal`.

I hope that the solution [proposed above](https://github.com/WordPress/gutenberg/pull/57255#discussion_r1442663814) can be a good starting point for a better UX

jorgefilipecosta replied:
I applied the patch suggested by @ciampo and implemented the remaining todo specified there. Thank you a lot for providing a detailed patch! Things seem to be working well.
The current behavior is:
- When an action without modal is executed the menu closes and the focus goes to bulk editor button.
- When an action opens a modal the menu keeps open if the modal is closed the focus goes to the button that triggered the modal, if the action in the modal is executed the menu gets closed and the focus goes back to the bulk edit button.

@youknowriad @ciampo  let me know if there is any other update required.

ciampo commented on packages/dataviews/src/bulk-actions.js:
```diff
@@ -0,0 +1,207 @@
+			key={ action.id }
+			disabled={ eligibleItems.length === 0 }
+			onClick={ async ( event ) => {
+				event.preventDefault();
```
DropdownMenuItem components by default cause the menu to close when clicked. If we want to change this behavior, we should use the `hideOnClick` prop instead of using `event.preventDefault()`.

ciampo replied:
The solution proposed in [this comment](https://github.com/WordPress/gutenberg/pull/57255#discussion_r1442663814) uses `hideOnClick` instead of `event.preventDefault()`

jorgefilipecosta replied:
Good one, I will look into other places and remove event.preventDefault(). I followed what we were doing in other places and I think hideOnClick may be a better fit there too.

jameskoster commented:
> I guess for me the answer to this is that I don't see bulk editing as the primary action on the lists, and something you'd rarely don, only for advanced use-cases

I'd agree bulk actions are not always primary, but I don't know that they're 'only advanced'. There are some fairly common flows where bulk actions are situationally primary, e.g. moderating comments, deleting inactive plugins, updating tags or categories. I'd still like to explore fine tuning the visual prominence before introducing a dedicated mode, there are many things to try there.

jorgefilipecosta committed:
fix flex shring

jorgefilipecosta committed:
Fix focus issues

jorgefilipecosta committed:
remove data items meanwhile removed from the selection

jorgefilipecosta committed:
remove line accidently added

youknowriad commented on packages/edit-site/src/components/actions/index.js:
```diff
@@ -33,7 +33,9 @@ export const trashPostAction = {
 	hideModalHeader: true,
-	RenderModal: ( { item: post, closeModal } ) => {
+	RenderModal: ( { items: posts, closeModal } ) => {
+		// Todo - handle multiple posts
```
Is this something we plan to address here?

jorgefilipecosta replied:
This PR is already big so I planned to implement bulk actions for pages in a follow-up.

youknowriad commented on packages/edit-site/src/components/page-pages/index.js:
```diff
@@ -339,6 +339,22 @@ export default function PagePages() {
 					view={ view }
 					onChangeView={ onChangeView }
 					onSelectionChange={ onSelectionChange }
+					labels={ {
```
Can we find a way to avoid this prop, find generic labels or something?

jorgefilipecosta replied:
The generic labels would need to be something like "Select a new item" or "Deselect item", we could not even reference the title because the dataviews component itself is not aware of what field is a "title". We already have this as a fallback if the labels are not passed.
Not describing the item you are selecting on the label seems like a11y regression compared to the current page PHP list in core. I think we will need this prop or at least a getItemTitle prop (similar to getItemId) that provides a title of the row so we could at least do something like "Select item: %s", where %s is the page or template title. The labels property is more descriptive and differentiates templates and pages but I would also be ok with a  getItemTitle prop and something more generic, let me know what option you would prefer.

youknowriad replied:
I think what bothers me the most is that these are functions, maybe we should keep the "sprintf" call within the component and just make these regular strings.

youknowriad replied:
in other words:

```
labels={ { selectItem: __( 'Select page: %s' ), deselectItem: __( 'Deselect page: %s' ) } }
```

jorgefilipecosta replied:
I applied the suggestion.

ntsekouras commented on packages/edit-site/src/components/page-templates/template-actions.js:
```diff
@@ -89,29 +112,82 @@ export const deleteTemplateAction = {
+								__(
+									'Are you sure you want to delete %s items?'
+								),
+								decodeEntities( templates.length )
```
No need for `decodeEntities` here. This also applies [above](https://github.com/WordPress/gutenberg/pull/57255/files#diff-a147bc579639fdc67bfc7fbc2b5a319cbb46f90a99430ac5c0005e58ce390dbdR64).

jorgefilipecosta replied:
Good catch 👍

ntsekouras commented on packages/edit-site/src/components/page-templates/template-actions.js:
```diff
@@ -36,34 +36,57 @@ export function useResetTemplateAction() {
-						template.type,
-						template.id
+					await Promise.all(
+						templates.map( ( template ) => {
```
I think it's probably better to perform the actions for each template iteratively in one loop.
That means for every item(1. revert, 2. save) and then move on to the next item.

Additionally for bulk actions it's better to use [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled). Promise all will stop if on any rejection which means previous fulfilled promises were successful. So while for example we bulk delete 3 items and the third is rejected, the other two has been deleted.

jorgefilipecosta replied:
Feedback applied 👍

jorgefilipecosta committed:
Fix vertical alignment post rebase.

jorgefilipecosta committed:
Fix unrequired decodeEntities usage.

jorgefilipecosta committed:
Apply feedback to the promises

mcsf commented on packages/edit-site/src/components/page-templates/template-actions.js:
```diff
@@ -36,34 +36,51 @@ export function useResetTemplateAction() {
+						templates[ 0 ].type === TEMPLATE_POST_TYPE
+							? _n(
+									'An error occurred while reverting the template.',
+									'An error occurred while reverting the templates.',
+									templates.length
 							  )
-							: __(
-									'An error occurred while reverting the template part.'
+							: _n(
+									'An error occurred while reverting the template part.',
+									'An error occurred while reverting the template parts.',
+									templates.length
```
It took me a while to wrap my head around it, but `_n` should not be used to distinguish "one item" from "more than one". It's more about grammatical plurality than arithmetic plurality — in many languages, these two coincide, but that's not always the case (e.g. Russian).

See:
* https://codex.wordpress.org/I18n_for_WordPress_Developers#Plurals
* https://developer.wordpress.org/reference/functions/_n/#comment-2397

If I understood properly, the right approach is:

```js
templates.length === 1
  ? __( 'An error occurred while reverting the template.' )
  : __( 'An error occurred while reverting the templates.' )
```

Note, however, that this other instance (in `bulk-actions.js`) is correct, because it's not treating "1" as a special case:

```js
_n(
  'Edit %d item',
  'Edit %d items',
  selection.length
),
```

jorgefilipecosta replied:
Very interesting, thank you for going deep on this 👍 Today I learned 😊

mcsf commented on packages/edit-site/src/components/page-templates/template-actions.js:
```diff
@@ -89,29 +106,82 @@ export const deleteTemplateAction = {
+									);
+									createSuccessNotice(
+										__(
+											'The selected items were deleted with success.'
```
```suggestion
											'Items deleted.'
```

mcsf commented on packages/edit-site/src/components/page-templates/template-actions.js:
```diff
@@ -89,29 +106,82 @@ export const deleteTemplateAction = {
+								// translators: %s: The template or template part's title.
+								__(
+									'Are you sure you want to delete %s items?'
+								),
```
See my lengthier comment about `_n`. Here, we should still provide a singular alternative for the locales that need it:

```js
templates.length > 1
  ? sprintf( _n( 'delete % item?', 'delete %s items?', templates.length ), templates.length )
  : sprintf( __( 'delete "%s"?' ), title )
```

mcsf commented on packages/edit-site/src/components/page-templates/template-actions.js:
```diff
@@ -89,29 +106,82 @@ export const deleteTemplateAction = {
+								// translators: %s: The template or template part's title.
+								__( 'Are you sure you want to delete "%s"?' ),
+								decodeEntities(
+									templates && templates[ 0 ]?.title?.rendered
```
```suggestion
									templates?.[ 0 ]?.title?.rendered
```

mcsf commented on packages/edit-site/src/components/page-templates/template-actions.js:
```diff
@@ -89,29 +106,82 @@ export const deleteTemplateAction = {
+										type: 'snackbar',
+									} );
+								}
+							} else {
+								await removeTemplate( templates[ 0 ], {
+									allowUndo: false,
+								} );
+							}
```
Why does the `templates.length > 1` branch need to do a lot of manual wrangling of entities, error-checking, etc., while the older branch has this convenient `removeTemplate` action? Can we make this more consistent and convenient?

jorgefilipecosta replied:
Hi @mcsf,
Because the removeTemplate takes care of error handling displaying the success notices etc... We only need to call it and everything is handled by the action.
We would have two ways to remove error handling from here:
-  Call removeTemplate in a loop. - The use experience would not be great instead a single notice saying x items were removed we would have a notice per item.
- Implement a removeTemplates action or expand removeTemplate to support multiple templates - Also not a great option we don't have any other case yet where we want to remove multiple templates so I'm not sure exposing an action for that is a good idea for now,

Given that the two approaches were not great I opted for doing the error handling inline.

mcsf replied:
Yeah, mine was a leading question, it was more about: can we take the time to create some symmetry to keep the codebase less surprising? For example:

* Create a new action, `removeTemplates`, but keep it private using the `private-apis` module.
* Refactor the preexisting `removeTemplate` to simply call the new one. We see this pattern across Gutenberg (`replaceBlock` -> `replaceBlocks`).
* The new `removeTemplates` should consider `templates.length` in the strings it generates for success and error notices.

I'm happy to be challenged on this suggestion, btw.

jorgefilipecosta replied:
I followed the suggestion and introduced a removeTemplates private API 👍

mcsf commented on packages/dataviews/src/view-table.js:
```diff
@@ -315,7 +388,13 @@ function ViewTable( {
+	onSelectionChange,
+	labels,
 } ) {
+	const hasBulkActions = useMemo( () => {
+		return actions?.some( ( action ) => action.supportsBulk );
+	}, [ actions ] );
```
Is `useMemo` really needed here?

* The result is a primitive value (boolean), so no need to reuse references
* The computation itself doesn't call any methods and should be immediate

Ah, maybe this is an artefact from adapting an older `actions?.filter( ... )` call?

jorgefilipecosta replied:
I have a tendency for when there is an iteration on array cache the result to avoid iterating on the array everytime the component rerenders but yah iterating on the array is not something costly so we can remove the cache.

jorgefilipecosta committed:
lint fix

jorgefilipecosta committed:
Feedback application

jorgefilipecosta committed:
Labels object instead of function

jorgefilipecosta committed:
Introduce removeTemplates action

jorgefilipecosta committed:
fix css rebase issue

oandregal commented on packages/dataviews/src/dataviews.js:
```diff
@@ -28,15 +29,34 @@ export default function DataViews( {
 	actions,
 	data,
 	getItemId = defaultGetItemId,
+	getItemTitle,
```
I feel this prop is field API territory: its purpose is telling dataviews what is the _field_ that we should use for announcing selection. It's doing what `getValue`/`render` should be doing. Knowing what's the "field that communicates the title" is something we need, and not only for bulk selection.

Could we make this explicit and model it as part of the field API? We could leverage the `primaryField`, for instance:

- make the table layout declare its primary field: it just happens to be `title` in pages & templates, but it doesn't have to.
- update the code that uses `getItemTitle` to use `primaryField.getValue` instead.

jorgefilipecosta replied:
Hi @oandregal, I updated the code to use primaryField.getValue, and removed getItemTitle.

oandregal commented on packages/dataviews/src/dataviews.js:
```diff
@@ -28,15 +29,34 @@ export default function DataViews( {
 	onSelectionChange = defaultOnSelectionChange,
 	onDetailsChange = null,
 	deferredRendering = false,
+	labels,
```
Can we provide default labels, such as `Deselect: (title)` or `Deselect item: (title)` so consumers don't _have_ to?

I'd even go further and don't introduce this property in this PR. We may want to later, but I'd like us to explore alternatives where consumers don't have to provide any labels at all to DataViews. The users already know _where_ they are (templates, pages, patterns) when they experience these labels (in a visual or auditive way). So, generic approaches such as "Deselect item" may work well. If we strongly wanted to do this, I agree a single object `labels` is the cleaner approach.

jorgefilipecosta replied:
We already provide defaults, if consumers don't provide the labels prop. The labels prop is optional if consumers want to provide better labels. I will follow the feedback and just have the generic version.

oandregal replied:
Ah, I saw the defaults now. What do you think of organizing them differently?

- The `DataViews` components prepares the `labels` in the same way it prepares the `fields`. Something along these lines (pseudocode):

```javascript
const _labels = useMemo(
  return {
    selectItem: labels.selectItem || __( 'Select a new item' ),
    deselectItem: labels.deselectItem || __( 'Deselect a new item' ),
  };
[ labels ] );

// ... 

<ViewComponent
  labels={ _labels }
/>
```

- The view component access the labels directly, because they are always present: the code of the view component is simpler because we can remove the conditionals. It also has the benefit of all views having the same labels.

jorgefilipecosta replied:
I ended up following your previous feedback and removed the labels property.

oandregal commented on packages/dataviews/src/dataviews.js:
```diff
@@ -28,15 +29,34 @@ export default function DataViews( {
+	useEffect( () => {
+		if (
+			selection.length > 0 &&
+			selection.some(
+				( id ) => ! data.some( ( item ) => item.id === id )
+			)
```
What would you think of rewriting this as:

```suggestion
			selection.some( ( id ) => ! data.includes( id ) )
```

oandregal replied:
Ah, data contains objects, not the IDs. We could still do (pseudocode):

```javascript
useEffect( () => {
  const dataIds = data.map( { id } => id );
  if ( selection.some( ( id ) => ! dataIds.includes( id ) ) {
    const newSelection = selection.filter( ( id ) => dataIds.includes( id );
  } 
}, [ selection, data ] );
```

Perhaps it's just my brain today, but I found concatenating two `some` conditions a bit confusing, and the semantics of `includes` much better for this case.

jorgefilipecosta replied:
Fixed 👍

jorgefilipecosta replied:
Hi @oandregal, I needed to revert this change, data.includes( id ) does not work because data is an array of templates and not an array of ids.

oandregal replied:
Yeah, I provided a second example that works. It's not a blocker, though.

oandregal commented on packages/dataviews/src/style.scss:
```diff
@@ -18,6 +18,10 @@
 	}
 }
 
+.dataviews__filters-view-actions.components-h-stack {
```
```suggestion
.dataviews-filters__view-actions.components-h-stack {
```

jorgefilipecosta replied:
Fixed 👍

oandregal commented on packages/dataviews/src/style.scss:
```diff
@@ -373,7 +377,23 @@
 	padding: 0 $grid-unit-40;
 }
 
+.dataviews-table-selection-checkbox label {
```
There was [a PR](https://github.com/WordPress/gutenberg/pull/57651) to do some consolidation to CSS naming. The current state is far from perfect, but could we follow the existing pattern in this one? It'd be:

```css
.dataviews-view-table-selection-checkbox label {
```

jorgefilipecosta replied:
Fixed 👍

oandregal commented:
If there are no actions available to perform, we could disable the bulk editing – the same way the inline actions button is.

https://github.com/WordPress/gutenberg/assets/583546/d363a770-ba53-4417-9316-d78ceeb0ee99

oandregal commented on packages/dataviews/src/view-table.js:
```diff
@@ -365,6 +452,22 @@ function ViewTable( {
 			>
 				<thead>
 					<tr>
+						{ !! selection && hasBulkActions && (
```
Is there any situation in which `!! selection` is different from `true`? If there is no selection, it's an empty array. Could this be just `hasBulkActions`?

If there were cases where selection could be false, we could use a `showBulkActions` bool instead (update it above):

```js
const showBulkActions = !! selection && actions?...
```

jorgefilipecosta replied:
Fixed 👍

oandregal commented on packages/dataviews/src/view-table.js:
```diff
@@ -306,16 +307,102 @@ function WithSeparators( { children } ) {
+	getItemId,
+	getItemTitle,
+} ) {
+	const id = getItemId?.( item );
```
Is there any situation in which `getItemId` may not be declared? Do we need to cover against that case? (saw this pattern a bunch of times: my understanding is that it's always defined, so we don't have to do this)

jorgefilipecosta replied:
We provide a default so we can remove the check and assume it is always declared.

jorgefilipecosta committed:
lint fixes.

jorgefilipecosta committed:
remove labels

jorgefilipecosta committed:
simplify logic condition

jorgefilipecosta committed:
fix some classnames

jameskoster commented:
> If there are no actions available to perform, we could disable the bulk editing – the same way the inline actions button is.

That gets a little tricky, because you'd still be able to select items via the checkboxes, but not do anything with them. I worry that might feel broken. If we wanted to try this, then it would probably make sense to disable the checkboxes for items that have no applicable actions as well.

At least with the menu being active we can include a tooltip to explain when an action cannot be applied to the current selection.

jorgefilipecosta committed:
multiple small changes of feedback

jorgefilipecosta committed:
remove getItemTitle

jorgefilipecosta commented:
> > If there are no actions available to perform, we could disable the bulk editing – the same way the inline actions button is.
> 
> That gets a little tricky, because you'd still be able to select items via the checkboxes, but not do anything with them. I worry that might feel broken. If we wanted to try this, then it would probably make sense to disable the checkboxes for items that have no applicable actions as well.
> 
> At least with the menu being active we can include a tooltip to explain when an action cannot be applied to the current selection.



I agree disabling the bulk edit may look broken in some cases. Having it enabled provides a way to select and deselect all, in most cases at least an action will be possible in that case.

oandregal commented:
> That gets a little tricky, because you'd still be able to select items via the checkboxes, but not do anything with them. I worry that might feel broken.

My point was to disable all of bulk editing (both the button at the top-right and the checkboxes).

I echo Riad's comment on bulk editing being too prominent. But more specially when it cannot be used, such as in the case I brought up. It's fine to ship as it is and iterate, though I feel this should be improved.

mcsf commented on packages/edit-site/src/store/private-actions.js:
```diff
@@ -49,3 +53,70 @@ export const setEditorCanvasContainerView =
+			let successMessage;
+
+			if ( templates.length === 1 ) {
+				// Depending on how the entity was retrieved it's title might be
```
```suggestion
				// Depending on how the entity was retrieved its title might be
```

jorgefilipecosta commented:
> > That gets a little tricky, because you'd still be able to select items via the checkboxes, but not do anything with them. I worry that might feel broken.
> 
> My point was to disable all of bulk editing (both the button at the top-right and the checkboxes).
> 
> I echo Riad's comment on bulk editing being too prominent. But more specially when it cannot be used, such as in the case I brought up. It's fine to ship as it is and iterate, though I feel this should be improved.

If bulk editing can not be used at all e.g: there are no bulk actions bulk edit does not appear. For example on pages or patterns we don't show bulk action controls yet.
Essently this check would be an additional one where we check if any of the bulk actions can be applied to any of items. It also has a performance cost of doing this check personally I'm not sure if it is worth it but I don't have a strong feeling and can add this additional check.

jameskoster commented:
> If bulk editing can not be used at all because there are no bulk actions bulk editing does not appear. For example, on pages or patterns, we don't show bulk action controls yet.

That sounds okay to me if it's part of the view config. IE if a view will _never_ have bulk actions, then the bulk actions UI should be hidden.

But for views like Templates where bulk actions are situationally available I think we should keep the UI visible and communicate via the menu when actions aren't available. Otherwise, the user is left to guess why bulk action controls are missing which can feel broken.

I'd still like to try reducing the _visual_ prominence of bulk actions. We can try smaller checkboxes, and only displaying them on hover/focus. Perhaps even a view option to hide the checkbox column.

jameskoster commented:
@jorgefilipecosta I pushed a couple of small style tweaks. If it's not too much work could we also try:

1. "Reset template", "Delete template" reads awkwardly when you have >1 item selected. These labels can be simplified to just "Reset", "Delete".
2. Could a class be added to the `tr` of selected rows so that we can visually highlight them like so:

<img width="1322" alt="Screenshot 2024-01-10 at 17 22 28" src="https://github.com/WordPress/gutenberg/assets/846565/b8e7ce3c-aa2a-4060-a0d8-c8ffa8711386">

jameskoster commented:
Also pushed a change to reduce the visual prominence of checkboxes. Let me know what y'all think, happy to tackle this separately if preferred. 


https://github.com/WordPress/gutenberg/assets/846565/e46e1fce-2a00-4bfe-8247-6d1aea5d7aae

ntsekouras commented:
>Would it be feasible to toggle the checkbox regardless of where you click in the tr? This could be a useful power-user tool that paves the way for cmd-click / shift-click multi-selection.

I think this is coupled with the inline editing interaction we want to achieve and with the fields that have clickable content(such as links). In order to make this happen I guess we would need to disable the inner interactions..

oandregal commented:
Gave it a second try:

https://github.com/WordPress/gutenberg/assets/583546/09544943-afda-4ab4-8aff-a3f0ef13a74a

I do see the bulk editing interface when no action is available to perform, but it is one of those situational states, as Jay mentioned: when I edit a template, then the _revert_ action becomes available. It makes sense.

With the styling tweaks, it carries a lot less visual weight and importance. Thanks for that.

oandregal reviewed (approved):
API-wise, this is ready: I like that we didn't have to introduce any new concept to DataViews to make bulk editing work.

Interaction-wise, this is a great starting place.

ntsekouras reviewed (changes requested):
@jorgefilipecosta we need to also update the [pattern actions ](https://github.com/WordPress/gutenberg/commits/trunk/packages/edit-site/src/components/page-patterns/dataviews-pattern-actions.js)with the new API to expect multiple items.

ntsekouras commented on packages/dataviews/src/view-table.js:
```diff
@@ -306,6 +307,80 @@ function WithSeparators( { children } ) {
 }
 
+function BulkSelectionCheckbox( { selection, onSelectionChange, data } ) {
+	const areAllSelected = selection && selection.length === data.length;
```
Nit: We always initiate selection with an empty array, so no need for the first check, right?

ntsekouras commented on packages/edit-site/src/store/private-actions.js:
```diff
@@ -49,3 +53,70 @@ export const setEditorCanvasContainerView =
+	( templates ) =>
+	async ( { registry } ) => {
+		try {
+			await Promise.allSettled(
```
The point of using `allSettled` is to inspect the return value which is of type:
```
[
  { status: 'fulfilled', value: 99 },
  { status: 'rejected', reason: Error: an error }
]
```

So the logic below for showing the messages is not right I think.. Let me check a bit better.

mcsf replied:
I agree with this. We should definitely be inspecting the results of `allSettled`.

ntsekouras replied:
Okay, after investigating more even if `deleteEntityRecord` is a thunk, the returned result is a Promise. My initial tests where wrong by throwing inside the `map` callback. 

So what we need to do here is just probably get the result of `allSettled` and inspect that.

jorgefilipecosta replied:
I updated the have the error logic based on allSettled result. It seems to be working well when I forced some delete errors on the backend.

mcsf commented on packages/edit-site/src/components/page-templates/template-actions.js:
```diff
@@ -89,29 +115,46 @@ export const deleteTemplateAction = {
+						? sprintf(
+								// translators: %d: number of items to delete.
+								_n(
+									'delete %d item?',
+									'delete %d items?',
```
```suggestion
									'Delete %d item?',
									'Delete %d items?',
```

mcsf commented on packages/edit-site/src/components/page-templates/template-actions.js:
```diff
@@ -89,29 +115,46 @@ export const deleteTemplateAction = {
+						  )
+						: sprintf(
+								// translators: %s: The template or template part's titles
+								__( 'delete "%s"?' ),
```
```suggestion
								__( 'Delete "%s"?' ),
```

jameskoster commented:
> I think this is coupled with the inline editing interaction we want to achieve and with the fields that have clickable content(such as links). In order to make this happen I guess we would need to disable the inner interactions..

My thinking was that this would separate. 'Inner interactions' would occur when you click the field values – presumably these would be interactive elements, likely `button`s inside each `td`. The click-tr-to-select would be a 'hidden' affordance for mouse users – no interactive element. Totally fine to handle this separately if needed.

Otherwise, in terms of design there are two pieces missing:

1. "Reset template" / "Delete template" can be simplified: "Reset" / "Delete"
2. When a record is selected, add a class to the `tr` so it can be visually highlighted. See the mockup in https://github.com/WordPress/gutenberg/pull/57255#issuecomment-1885287175.

jorgefilipecosta committed:
Inspect Promise.allSettled result

jorgefilipecosta committed:
typo

jorgefilipecosta committed:
case fixing

jorgefilipecosta committed:
update patterns actions

jorgefilipecosta committed:
Added a selection mark

jorgefilipecosta committed:
lint fixes

jameskoster commented:
@jorgefilipecosta my style changes from yesterday seem to have vanished 🤔 Is there any way to retrieve them or must I re-do that work?

jameskoster committed:
Style adjustments

jorgefilipecosta commented:
> @jorgefilipecosta my style changes from yesterday seem to have vanished 🤔 Is there any way to retrieve them or must I re-do that work?
> 
> Edit: Never mind, I re-did it :d

I'm sorry I'm not sure what happened probably some rebase issue 😞 thank you for re-adding the commit 🙇

jorgefilipecosta commented:
All the blockers have been addressed thank you all for the feedback, reviews, commits, and suggestions. I will merge this PR so we can continue with follow-ups and related work and avoid other big rebases. However, any feedback is still welcome and will be applied as a follow-up.
